### PR TITLE
fix: Use `makeCopy` to change relation in `FileSourceScanExec`

### LINKED ISSUE #190: java.lang.NoSuchMethodError: org.apache.spark.sql.execution.FileSourceScanExec.copy$default$9()
### Describe the bug

I want to try Comet on Databricks.
I built Comet for Spark 3.4 (`-Prelease -Pspark-3.4`), added it to the jars via init-script, added all needed configurations (`spark conf`). When I am running any query on delta-tables, it tends to the following:
```
com.databricks.backend.common.rpc.SparkDriverExceptions$SQLExecutionException: java.lang.NoSuchMethodError: org.apache.spark.sql.execution.FileSourceScanExec.copy$default$9()Z
	at org.apache.spark.sql.comet.CometScanExec$.apply(CometScanExec.scala:442)
	at org.apache.comet.CometSparkSessionExtensions$CometScanRule$$anonfun$apply$5.applyOrElse(CometSparkSessionExtensions.scala:127)
	at org.apache.comet.CometSparkSessionExtensions$CometScanRule$$anonfun$apply$5.applyOrElse(CometSparkSessionExtensions.scala:77)
	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDownWithPruning$1(TreeNode.scala:465)
	at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(origin.scala:69)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDownWithPruning(TreeNode.scala:465)
	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDownWithPruning$3(TreeNode.scala:470)
	at org.apache.spark.sql.catalyst.trees.UnaryLike.mapChildren(TreeNode.scala:1269)
	at org.apache.spark.sql.catalyst.trees.UnaryLike.mapChildren$(TreeNode.scala:1268)
	at org.apache.spark.sql.execution.FilterExec.mapChildren(basicPhysicalOperators.scala:216)
```

### Steps to reproduce

1. Built Comet for spark 3.4
2. Add init-script to Databricks cluster, that adds Comet-JAR to `/databricks/jars`
3. Run any query on delta-tables

### Expected behavior

Comet is working on Databricks 13.3 LTS (Spark 3.4.1, Scala 2.12). Comet fallback to delta-io for reading delta-tables maybe?

### Additional context

I can provide more extended stack-trace, but there is nothing special before lines above, only catalyst-calls.

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Relates #190.

### CODE REVIEW

**Comment by viirya** [comment]
> mapProductIterator in class TreeNode cannot be accessed in org.apache.spark.sql.execution.FileSourceScanExec

Ah, `mapProductIterator` is protected method. Let me update this later.

File: spark/src/main/scala/org/apache/spark/sql/comet/CometScanExec.scala [hunk 1]
```diff
@@ -439,8 +440,29 @@ case class CometScanExec(
 
 object CometScanExec {
```
Comment (comphead) [inline spark/src/main/scala/org/apache/spark/sql/comet/CometScanExec.scala:441]: perhaps we can also consider `CometScanExec extends TreeNode` to remove `mapProductIterator` method copy?

Comment (viirya) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometScanExec.scala:441]: We want to run `mapProductIterator` on `FileSourceScanExec`. I'm not sure how `object CometScanExec extends TreeNode` can help us?

**Comment by viirya** [comment]
This patch cannot completely resolve the ticket #190 but only fixes the first error. After the first error is fixed, there comes out another API incompatibility issue around `DataSourceScanExec`. 

I think it is still good to have this patch to fix the first error.

File: spark/src/main/scala/org/apache/spark/sql/comet/CometScanExec.scala (hunk 1 above)
Comment (comphead) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometScanExec.scala:441]: You are right, its needed to run `mapProductIterator` on `FileSourceScanExec`, I wrongly assumed this method needed for `CometScanExec `.

**Review by comphead — APPROVED** [review]
lgtm thanks @viirya

**Comment by viirya** [comment]
Merged. Thanks.

**Pull request merged**